### ISSUE TITLE
Ability to replicate/recompute tensors.

### ISSUE THREAD

**Issue by csarofeen** [csarofeen/pytorch#139]
From the computeAt work, one thing we assume right now is that all tensors will only be produced once. This is typically a good default, but does not support everything we may need to do. For example if we can have persistent softmax, this would be fine, but non-persistent softmax would require recomputing some values.

We need a mechanism where we can specify the use of a tensor in an expression, and replicate expressions from its origin to inputs, so that it will be recomputed, with an interface that returns all recomputed tensors.

This will be functionality that we will add to computeAt, so if computeAt recognizes it can't satisfy a given computeAt call it can have an option for the user where it will automatically recompute all conflicting tensors.

**Comment by naoyam** [comment]
A related comment: https://github.com/csarofeen/pytorch/pull/440/files#r520620322

**Comment by csarofeen** [comment]
This is implemented now in https://github.com/csarofeen/pytorch/blob/devel/torch/csrc/jit/codegen/cuda/ir_cloner.h#L105
